### ISSUE TITLE
run.command() interface & path quoting

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#1845]
Branching discussion from #1689, specifically [this comment](https://github.com/MRtrix3/mrtrix3/pull/1689#issuecomment-564711729), so as not to blur distinct issues.

> As you can see, the issue is that the argument passed to mrconvert's -copy_properties was quoted with single quotes (presumably by the `shlex.quote()` call), but these are then passed along with the argument itself to the command. I can fix this by adding an additional `.strip ("'")` to this line in `run.py`, so that it strips both single and double quotes.

I have never previously seen single-quotes introduced like that for the purpose of encapsulating filesystem paths. I would need you to trace back and find precisely what the source of it is, so that it can either be modified to be double-quote at the source, or all API filesystem path handling checked to ensure compatibility with either single or double quotes.

> This was kind of discussed in #928, but I do think that things would really be a lot easier if we passed lists of strings to the `run.command()` call rather than trying to tokenise a single command string, with all the gotchas involved in doing this...

Of course it'd be easier from an API management perspective to do so; but it'd be way less convenient when actually writing scripts to run against the API. Constructing command strings to pass to `run.command()` is a lot more intuitive than constructing `argv`s; hence why I've put so much effort into providing that capability for my own sake. So removing that capability would feel like a pretty large regression to me. Unless it becomes unmanageable, I'd rather put in the effort to keep string-based inputs to `run.command()` working. If it's just a matter of a particular version of `shlex.quote()` using singles rather than doubles, it shouldn't be too hard to rectify; much easier than what's already been put in place in this context.

**Comment by jdtournier** [comment]
Yes, I agree it would be quite a bit of effort and a loss in convenience if we had to enforce the use of lists instead of command strings - not to mention throwing out quite a bit of work you've done on this. But it would avoid these types of issues...

In any case, I think I've figured out where it's coming from, and the fix is simple - see c7f91da. I have a feeling this will need to be replicated across all instances where the `mrconvert_keyval` argument is set in `run.command()`... It might be simpler to allow that command to take the original input (not the output of `path.from_user()`), which `run.command()` could then handle appropriately internally. I guess this would mean it couldn't be used for cases where `app.WORKING_DIR` is not set? Or maybe that's not an issue? The only instances I could identify where `mrconvert_keyval` doesn't receive the output of `path.from_user()` are these ones (which I'm not sure I understand, TBH):
```
bin/dwifslpreproc:  run.command('mrconvert result.mif ' + path.from_user(app.ARGS.output) + grad_export_option, mrconvert_keyval='output.json', force=app.FORCE_OVERWRITE)
bin/dwicat:  run.command('mrconvert result.mif ' + path.from_user(app.ARGS.output), mrconvert_keyval='result_final.json', force=app.FORCE_OVERWRITE)
lib/mrtrix3/dwinormalise/group.py:  run.command('mrconvert template_wm_mask.mif ' + path.from_user(app.ARGS.wm_mask), mrconvert_keyval='NULL', force=app.FORCE_OVERWRITE) 
lib/mrtrix3/dwinormalise/group.py:  run.command('mrconvert fa_template.mif ' + path.from_user(app.ARGS.fa_template), mrconvert_keyval='NULL', force=app.FORCE_OVERWRITE) 
```
In any case, it can be fixed, but it might be nice to find a cleaner solution that it's likely to lead to problems down the track when people forget about this and don't add that extra `False` at the end of their `path.from_user()` call...

**Comment by Lestropie** [comment]
> I have a feeling this will need to be replicated across all instances where the `mrconvert_keyval` argument is set in `run.command()`

Correct: it will need to be `path.from_user(escape=False)` whenever referencing a user input image.

> It might be simpler to allow that command to take the original input (not the output of `path.from_user()`), which `run.command()` could then handle appropriately internally. I guess this would mean it couldn't be used for cases where `app.WORKING_DIR` is not set?

The problem here is that sometimes key-value data are imported not from a user-specified image (which will be referenced with respect to the original working directory), but from a JSON file explicitly constructed *within the scratch directory* (the `dwifslpreproc` and `dwicat` examples you found). One *could* do a file search in both prospective locations, but there's a plausible risk of finding files with that name in both locations. Better to:

1. Ensure that for any instance where the input to `mrconvert_keyval` is based on `path.from_user()`, `escape=False` is specified.

2. In `run.command()`, remove any quotation marks from the `mrconvert_keyval` string - maybe by calling a new `path.unquote()` function - just in case a script author makes that mistake.

-----

> More generally, there is a problem that:

>      `quote (quote ('some problematic string'))`

> will not match:

>      `quote ('some problematic string')`

> since it'll try to quote whatever additions the previous invocation did
to properly quote the original string. I'm not sure whether there's any
way around that...

I'm surprised that `shlex.quote()` or `pipes.quote()` would behave like that. If that's the case, then we may want to make `path.quote()` a little more advanced, by invoking `path.unquote()` before runing the library `quote()` function.

-----

Given there's API changes necessary here, I'd rather cherry-pick c7f91da6 over to a new branch, make the remaining API changes, merge to `dev`, then import those changes into #1689.

**Comment by jdtournier** [comment]
I had another quick look into this, and I don't think this is going to be a simple fix. The issue is that the `mrconvert_keyval` argument is expected to be a single, _unquoted_ argument [at this point](https://github.com/MRtrix3/mrtrix3/blob/dwi2response_dhollander_version_switch/lib/mrtrix3/run.py#L315) in `run.command()`, since it's being appended to the command _list_. But it will typically have been processed through `path.from_user()`, which will by default provide a _quoted_ version of that string. The only fix I can think of is what's already in #1689: add the `escape=False` flag to `path.from_user()`. 

> we may want to make path.quote() a little more advanced, by invoking path.unquote() before runing the library quote() function.

I've had a look into that, and I just can't see how we can robustly do this. Maybe you can think of something clever here...
